### ISSUE TITLE
How to confirm every sample K value ？

### ISSUE THREAD

**Issue by Phoenix12580** [carmonalab/GeneNMF#19]
How to confirm every sample K value like this？  Thank you very much
![image](https://github.com/user-attachments/assets/5fdf253d-8179-4449-ae2b-269a2a4b2fb7)

**Comment by mass-a** [comment]
Hi, what is "coefficient" in this graph?

**Comment by Phoenix12580** [comment]
Thank you for your reply!

The cophenetic correlation coefficients were calculated to measure the stability and accordingly, a factorization rank of 5 was selected ([Figure S3](https://www.cell.com/cancer-cell/fulltext/S1535-6108(24)00234-4?_returnURL=https%3A%2F%2Flinkinghub.elsevier.com%2Fretrieve%2Fpii%2FS1535610824002344%3Fshowall%3Dtrue#mmc1)C).

https://www.cell.com/cancer-cell/fulltext/S1535-6108(24)00234-4?_returnURL=https%3A%2F%2Flinkinghub.elsevier.com%2Fretrieve%2Fpii%2FS1535610824002344%3Fshowall%3Dtrue  
I get this from this papaer

**Comment by mass-a** [comment]
Thanks! I wasn't aware of this use of the cophenetic coefficient.

I am not sure how this can be used within GeneNMF though. The algorithm runs NMF for multiple values of k, and then uses the results of **all** values of k, across multiple samples, to derive a consensus set of meta-programs. Determining the number of meta-programs (nMP) is somewhat subjective (e.g. what overlap do we want to allow between meta-programs), and we don't yet have robust quantitative criteria for the choice of nMP.

**Comment by Phoenix12580** [comment]
Thank you for your reply! I hope this article can provide ideas for you to improve GeneNMF！ 

I'm not sure if my approach is correct, but if we could know the exact K value for each sample that best represents the NMF program at best K value for that particular sample, and then perform a metaprogram analysis of the most representative NMF program from each sample, it might be more accurate.

**Comment by mass-a** [comment]
My thinking has been that, if a gene program is robust, we should observe it across multiple values of `k` (and possibly in more than one sample). So a scan across values of `k` should provide more reliable and conserved gene modules, what do you think?
But I like your idea, we could give it a try.

**Comment by Phoenix12580** [comment]
Yes, I do agree with you ! just like in the pan-cancer intra-tumor heterogeneity (ITH) article. However, considering that many researchers do not have access to a large number of samples, and given the strong heterogeneity of tumors, using this method to screen NMF programs with only a few samples introduces a significant degree of subjectivity and interpretability issues. In my previous practice, integrating more than 20 prostate cancer epithelial cell samples to select   meta program did not effectively remove batch effects, and the subsequent clustering results were also poor. Therefore, I wondered if this approach could be applied to small sample sizes and highly heterogeneous cancer types . The idea would be to first identify the best NMF programs for each sample based on their best K values, and then the NMF programs that are not integrated into the meta program might represent sample-specific programs, while the meta program would represent the common programs across this type of cancer.
